### PR TITLE
Prevent Float value from automatically becoming BigDecimal when using Oj

### LINKED ISSUE #67: Float with many digits automatically becomes BigDecimal when using Oj
# Detail

If a document has `coordinates` field with `[131.01468600332737, 43.0554632]`, then call `Elastic::Transport::Client#perform_request` like the following PoC.

```ruby
# poc.rb

require 'elastic/transport'
require 'oj'

client = Elastic::Transport::Client.new(hosts: { host: "localhost", port: 9200})
body = {
  from: 0,
  size: 1,
  query: {
    match: {
      key: "5yc1x"
    }
  }
}

response = client.perform_request('GET', 'development/_search', {}, body)

coordinates = response.body['hits']['hits'][0]['_source']['coordinates']
p coordinates
p coordinates.map(&:class)
```

```
% bundle exec ruby poc.rb
[0.13101468600332737e3, 43.0554632]
[BigDecimal, Float]
```

But if remove `require 'oj'`, then

```
% bundle exec ruby poc.rb
[131.01468600332737, 43.0554632]
[Float, Float]
```

This difference depends on the [mode](https://github.com/ohler55/oj/blob/develop/pages/Modes.md) of oj used by [MultiJson](https://github.com/intridea/multi_json). MultiJson loads JSON with oj as `strict` mode by default.

```
> num_str = "1.11111111111111111"

> MultiJson.load(num_str).class
BigDecimal < Numeric

> Oj.load(num_str, {mode: :strict}).class
BigDecimal < Numeric

> Oj.load(num_str, {mode: :null}).class
BigDecimal < Numeric

> Oj.load(num_str, {mode: :compat}).class
Float < Numeric

> Oj.load(num_str, {mode: :rails}).class
BigDecimal < Numeric

> Oj.load(num_str, {mode: :object}).class
BigDecimal < Numeric

> Oj.load(num_str, {mode: :custom}).class
BigDecimal < Numeric
```

MultiJson allows other [JSON engines](https://github.com/intridea/multi_json#supported-json-engines) to be selected, while all others remain Float.

# What kind do we have trouble?

For example, when returning JSON value in an API endpoint, it is not possible to specify either a numeric or a string as the value type. Therefore, unintended type conversion is required only when using oj.

```
> Oj.load(num_str, {mode: :strict}).as_json
"1.11111111111111111"
> Oj.load(num_str, {mode: :strict}).as_json.class
String < Object

> Oj.load(num_str, {mode: :compat}).as_json
1.1111111111111112
> Oj.load(num_str, {mode: :compat}).as_json.class
Float < Numeric
```

### PR DESCRIPTION
**Pull request by kyoshidajp**
Fixes #67

Only `compat` mode cat prevent automatically becoming BigDecimal when using oj.

### CODE REVIEW

File: lib/elastic/transport/transport/base.rb [hunk 1]
```diff
@@ -346,10 +346,17 @@ def perform_request(method, path, params = {}, body = nil, headers = nil, opts =
...
-                                                       response.headers &&
-                                                       response.headers["content-type"] =~ /json/
+          if response.body &&
+            !response.body.empty? &&
+            response.headers &&
+            response.headers["content-type"] =~ /json/
+
+            # Prevent Float value from automatically becoming BigDecimal when using Oj
+            options = {}
+            options[:mode] = :compat if serializer.is_a?(Serializer::MultiJson) && defined?(::Oj)
```
Comment (kyoshidajp) [inline lib/elastic/transport/transport/base.rb:356]: This may be a technical debt, but I couldn't figure out a better way.

**Comment by picandocodigo** [comment]
Thank you very much for this fix @kyoshidajp. Could you add a test for the bug please? Something like the code you added to the very detailed issue you reported would be very useful. Thanks!

**Comment by kyoshidajp** [comment]
@picandocodigo Thanks. I added a test.

**Comment by picandocodigo** [comment]
Thank you very much @kyoshidajp!

**Pull request merged**